asyncio_run_in_process/_child.py

```diff
@@ -70,11 +72,98 @@ def update_state_finished(to_parent: BinaryIO, finished_payload: bytes) -> None:
     to_parent.flush()
 
 
+async def _do_task_cleanup(*tasks: 'asyncio.Future[Any]') -> None:
+    for task in tasks:
+        if not task.done():
+            task.cancel()
+
+        try:
+            await task
+        except asyncio.CancelledError:
+            pass
+
+
 SHUTDOWN_SIGNALS = {signal.SIGTERM}
 
 
+async def _handle_SIGTERM(task: 'asyncio.Task[TReturn]', fut: 'asyncio.Future[int]') -> None:
+    signum = await fut
+    raise SystemExit(signum)
+
+
+async def _handle_coro(coro: Coroutine[Any, Any, TReturn], got_SIGINT: asyncio.Event) -> TReturn:
+    """
+    Understanding this function requires some detailed knowledge of how
+    coroutines work.
+
+    The goal here is to run a coroutine function and wait for the result.
+    However, if a SIGINT signal is received then we want to inject a
+    `KeyboardInterrupt` exception into the running coroutine.
+
+    Some additional nuance:
+
+    - The `SIGINT` signal can happen multiple times and each time we want to
+      throw a `KeyboardInterrupt` into the running coroutine which may choose to
+      ignore the exception and continue executing.
+    - When the `KeyboardInterrupt` hits the coroutine it can return a value
+      which is sent using a `StopIteration` exception.  We treat this as the
+      return value of the coroutine.
+
+    The raising and silencing of the `RuntimeError` is due to the coroutine
+    being wrapped in a `Task`.  At this point asyncio considers itself
+    responsible for the lifecycle of the coroutine since asyncio won't let you
+    abandon tasks.  This leaves us with a dilema.  If we don't `await` the task
+    the asyncio will complain that the task result/exception was never
+    retrieved.  But if we do `await` it in the case that we've thrown a
+    `KeyboardInterrupt` into the running coroutine, then asyncio complains that
+    it has already been awaited by raising a `RuntimeError`.  This leaves us
+    with the lesser of two evils, catching and silencing the `RuntimeError`
+    rather than leaving an abandoned task which would cause a warning to be
+    issued when the python process exits.
+    """
+    coro_task = asyncio.ensure_future(coro)
+    while True:
+        # Run the coroutine until it either returns, or a SIGINT is received.
+        # This is done in a loop because the function *could* choose to ignore
+        # the `KeyboardInterrupt` and continue processing, in which case we
+        # reset the signal and resume waiting.
+        done, pending = await asyncio.wait(
```

Comment:
Oh! After looking at the code for `asyncio.wait()` I think there might be a race-condition here. Even though this uses `FIRST_COMPLETED` it's possible for both of these futures to trigger; if both futures complete during the same pass of the event loop. If this happens then the call to `coro.throw(KeyboardInterrupt)` will throw a `RuntimeError`, because the coro has already finished. I think the fix is to guard that `throw` with a check to `coro_task.done()`.